From df76d1d1fdc26f9bf91248287becb3e13f122863 Mon Sep 17 00:00:00 2001
From: Raghavendra Kakarla <rkakarla@codeaurora.org>
Date: Sun, 11 May 2025 10:40:18 +0500
Subject: [PATCH] BACKPORT: driver: lpm-levels: Add check for IPI pendning in
 core lpm path Add check whether IPI is in pending state for the core while
 its entering to LPM.

Change-Id: I6927fd445a1e2e46a4b5329003994a5bd710dc66
Signed-off-by: Raghavendra Kakarla <rkakarla@codeaurora.org>
---
 drivers/cpuidle/lpm-levels.c               |  3 ++-
 include/trace/events/trace_msm_low_power.h | 17 +++++++++++++++++
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/drivers/cpuidle/lpm-levels.c b/drivers/cpuidle/lpm-levels.c
index b18f9cc75b5f..79163f62283a 100644
--- a/drivers/cpuidle/lpm-levels.c
+++ b/drivers/cpuidle/lpm-levels.c
@@ -1420,6 +1420,7 @@ static void update_ipi_history(int cpu)
 	if (history->current_ptr >= MAXSAMPLES)
 		history->current_ptr = 0;
 	history->cpu_idle_resched_ts = now;
+	trace_ipi_wakeup_time(ktime_to_us(now));
 }
 
 static void update_history(struct cpuidle_device *dev, int idx)
@@ -1469,7 +1470,7 @@ static int lpm_cpuidle_enter(struct cpuidle_device *dev,
 	cluster_prepare(cpu->parent, cpumask, idx, true, start_time);
 
 	trace_cpu_idle_enter(idx);
-	lpm_stats_cpu_enter(idx, start_time);
+	if (need_resched() || is_IPI_pending(cpumask_of(dev->cpu)))
 
 	if (need_resched())
 		goto exit;
diff --git a/include/trace/events/trace_msm_low_power.h b/include/trace/events/trace_msm_low_power.h
index 4e65444aaf54..3cba0ab02363 100644
--- a/include/trace/events/trace_msm_low_power.h
+++ b/include/trace/events/trace_msm_low_power.h
@@ -244,6 +244,23 @@ TRACE_EVENT(cluster_pred_hist,
 		__entry->sample, __entry->tmr)
 );
 
+TRACE_EVENT(ipi_wakeup_time,
+
+	TP_PROTO(u64 wakeup),
+
+	TP_ARGS(wakeup),
+
+	TP_STRUCT__entry(
+		__field(u64, wakeup)
+	),
+
+	TP_fast_assign(
+		__entry->wakeup = wakeup;
+	),
+
+	TP_printk("wakeup:%llu", __entry->wakeup)
+);
+
 #endif
 #define TRACE_INCLUDE_FILE trace_msm_low_power
 #include <trace/define_trace.h>
-- 
2.49.0

